homeassistant/components/evohome.py

```diff
@@ -0,0 +1,667 @@
+zones (e.g. TRVs, relays) and, optionally, a DHW controller.
+
+A minimal configuration.yaml is as as below:
+
+evohome:
+  username: !secret evohome_username
+  password: !secret evohome_password
+
+These config parameters are presented with their default values:
+
+# scan_interval: 180     # seconds, you can probably get away with 60
+# location_idx: 0        # if you have more than 1 location, use this
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/evohome/
+"""
+
+# Glossary
+# TCS - temperature control system (a.k.a. Controller, Master), which can
+# have up to 13 Slaves:
+#   0-12 Heating zones (a.k.a. Zone), and
+#   0-1 DHW controller, (a.k.a. Boiler)
+
+import logging
+from datetime import datetime, timedelta
+import requests
+import voluptuous as vol
+
+from homeassistant.components.climate import (
+    ClimateDevice,
+
+    SUPPORT_OPERATION_MODE,
+    SUPPORT_AWAY_MODE,
+)
+
+from homeassistant.const import (
+    CONF_USERNAME,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+
+    EVENT_HOMEASSISTANT_START,
+    PRECISION_TENTHS,
+
+    STATE_OFF,
+    STATE_ON,
+    TEMP_CELSIUS,
+
+    HTTP_BAD_REQUEST,
+    HTTP_TOO_MANY_REQUESTS,
+    HTTP_SERVICE_UNAVAILABLE,
+)
+
+# These are HTTP codes commonly seen with this component
+#   HTTP_BAD_REQUEST = 400          # usually, bad user credentials
+#   HTTP_TOO_MANY_REQUESTS = 429    # usually, api limit exceeded
+#   HTTP_SERVICE_UNAVAILABLE = 503  # this is common with Honeywell's websites
+
+from homeassistant.core import callback
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['evohomeclient==0.2.7']
+
+_LOGGER = logging.getLogger(__name__)
+
+# Usually, only the controller does client api I/O during update() to pull
+# current state - the exception is when zones pull their own schedules.
+# However, any entity can call methods that will (eventually) change state.
+PARALLEL_UPDATES = 0
+
+# these are specific to this component
+ATTR_UNTIL = 'until'
+
+DOMAIN = 'evohome'
+DATA_EVOHOME = 'data_' + DOMAIN
+DISPATCHER_EVOHOME = 'dispatcher_' + DOMAIN
+
+MIN_TEMP = 5
+MAX_TEMP = 35
+MIN_SCAN_INTERVAL = 180
+
+CONF_LOCATION_IDX = 'location_idx'
+
+# Validation of the user's configuration.
+CV_FLOAT = vol.All(vol.Coerce(float), vol.Range(min=MIN_TEMP, max=MAX_TEMP))
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL,
+                     default=MIN_SCAN_INTERVAL): cv.positive_int,
+
+        vol.Optional(CONF_LOCATION_IDX, default=0): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+# these are for the controller's opmode/state and the zone's state
+EVO_RESET = 'AutoWithReset'
```

Comment:
Please see https://github.com/home-assistant/architecture/issues/22. Are all of these possible states? We can't just have different states for a single climate platform

Reply:
I have previously read (and contributed to) that conversation.  Yes, they are all possible states, as reported by the client api.

The minutae is a little less simple than that, and I can do through the detail if you wish.

I have removed some CONSTANTS that applies only to the zones.

Reply:
@jeradM AM I OK to leave it as is?

BTW, this may give you an idea of the nature of it: http://www.automatedhome.co.uk/wp-content/uploads/2014/02/evohome-controller-display-home-en.png

Reply:
Regarding the above graphic, I should have said that the quick actions are the operating modes and the overriding decayed is that the state.

Reply:
We will not accept any PR that does not use the states provided by the climate component.

Because not using the built-in states means it doesn't work with Alexa, Google Assistant or any other system that will integrate with us, like Web of Things in the future.

Reply:
I am now using HA states.